WEEK 2 ASSIGNMENT
====
Frequent Itemsets and Association Rule

In [1]:
import pandas as pd
df = pd.read_csv('Online Retail.csv', encoding = "ISO-8859-1")
s2 =df.groupby('InvoiceNo')['StockCode'].apply(','.join)
s2 = s2.reset_index()
s2[:5].to_csv('Online Retail.txt', columns = ["StockCode"], header=False, index=False)

In [15]:
data = sc.textFile("Online Retail.txt")
transactions = data.map(lambda line: line.strip().split(' '))

In [38]:
model = FPGrowth.train(transactions, minSupport = 0.15, numPartitions = 10)
freqItemsets = model.freqItemsets().collect()

freqItemsetsDict = {}

for i in range(0, len(freqItemsets)):
    freqItemsetsDict[str(freqItemsets[i].items)] = freqItemsets[i].freq

1) How many frequent itemsets were found with minimum support of .15?

In [39]:
freqItemsets

[FreqItemset(items=['21756'], freq=1),
 FreqItemset(items=['"22633,22632"'], freq=1),
 FreqItemset(items=['"22960,22913,22912,22914"'], freq=1),
 FreqItemset(items=['"85123A,71053,84406B,84029G,84029E,22752,21730"'], freq=1),
 FreqItemset(items=['"84879,22745,22748,22749,22310,84969,22623,22622,21754,21755,21777,48187"'], freq=1)]

In [34]:
minConfidence = .4
rules = sorted(model._java_model.generateAssociationRules(minConfidence).collect(), 
    key=lambda x: x.confidence(), reverse=True)

for j in range(0, len(rules)):
    antecedent = list(rules[j].antecedent())
    consequent = list(rules[j].consequent())
    ruleItemset = str(antecedent + consequent)
    
    try:
        support = "{0:.15f}".format(freqItemsetsDict[ruleItemset] / transactions.count())
    except KeyError:
        next
    
    confidence = "{0:.15f}".format(rules[j].confidence())
    
    try:
        print(antecedent, '=>', consequent, '(support: '+ str(support), 'confidence: '+ str(confidence) + ')')
    except NameError:
        next

2) How many association rules were identified with a minimum confidence of .4?

As per the above code, the result is 0


3) What is the support and confidence of the following association rule: ['21754', '22748'] => ['22745']? Go beyond reporting the values, but include an interpretation of these evaluation metrics.

Support = 1/5 or .2
Confidence = 1/5 or .2 as well.

Logic: Since this is such a small dataset, the 5 transactions don't have any association rule. This data shows us that out of the 5 transactions there are no valuable conclusions that can be drawn. 